{{ cookiecutter.package_name|slugify }}/{% if cookiecutter.continuous_integration == "GitLab" %}.gitlab-ci.yml{% endif %}

```diff
@@ -13,48 +22,49 @@ stages:
     DOCKER_REGISTRY_USER: $CI_REGISTRY_USER
     DOCKER_REGISTRY_PASSWORD: $CI_REGISTRY_PASSWORD
   script:
-    - DOCKER_IMAGE_SHA="${DOCKER_IMAGE_SHA:-$(sha1sum Dockerfile poetry.lock pyproject.toml | sha1sum | cut -c 1-12)}"
-    - echo "CI_IMAGE_SHA=$DOCKER_IMAGE_SHA" >> .env
     - |
       echo "$DOCKER_REGISTRY_PASSWORD" | docker login --username "$DOCKER_REGISTRY_USER" --password-stdin "$DOCKER_REGISTRY"
       DOCKER_PUSH=${DOCKER_PUSH:-$(timeout 2s docker pull "$DOCKER_IMAGE":"$DOCKER_IMAGE_SHA" >/dev/null 2>&1 && echo $? || echo $?)}
-      if [ "$DOCKER_PUSH" -ne 1 ]; then
-        echo "$DOCKER_IMAGE:$DOCKER_IMAGE_SHA exists, skipping this job..."
-      else
-        # Compile a list of image tags consisting of a hash of its contents, the latest tag if this
-        # pipeline is running for the default branch, and the Git tag if this commit is tagged.
-        DOCKER_TAGS="$DOCKER_IMAGE_SHA"
-        if [ "$CI_COMMIT_BRANCH" = "$CI_DEFAULT_BRANCH" ]; then DOCKER_TAGS="$DOCKER_TAGS latest"; fi
-        if [ -n "$CI_COMMIT_TAG" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_COMMIT_TAG"; fi
-        if [ -n "$CI_ENVIRONMENT_NAME" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_ENVIRONMENT_NAME"; fi
-        DOCKER_TAGS_JOINED=""
-        for DOCKER_TAG in $DOCKER_TAGS; do
-          DOCKER_TAGS_JOINED="$DOCKER_TAGS_JOINED --tag $DOCKER_IMAGE:$DOCKER_TAG"
-        done
 
-        # Build the Docker image with all of the selected tags.
+      # Compile a list of image tags consisting of a hash of its contents, the latest tag if this
+      # pipeline is running for the default branch, and the Git tag if this commit is tagged.
+      DOCKER_TAGS="$DOCKER_IMAGE_SHA"
+      if [ "$CI_COMMIT_BRANCH" = "$CI_DEFAULT_BRANCH" ]; then DOCKER_TAGS="$DOCKER_TAGS latest"; fi
+      if [ -n "$CI_COMMIT_TAG" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_COMMIT_TAG"; fi
+      if [ -n "$CI_ENVIRONMENT_NAME" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_ENVIRONMENT_NAME"; fi
+      DOCKER_TAGS_JOINED=""
+      for DOCKER_TAG in $DOCKER_TAGS; do
+        DOCKER_TAGS_JOINED="$DOCKER_TAGS_JOINED --tag $DOCKER_IMAGE:$DOCKER_TAG"
+      done
+
+      # Build the Docker image with all of the selected tags.
+      {%- if cookiecutter.private_package_repository_name %}
+      echo "[http-basic.{{ cookiecutter.private_package_repository_name|slugify }}]" >> auth.toml
+      echo "username = \"gitlab-ci-token\"" >> auth.toml
+      echo "password = \"$CI_JOB_TOKEN\"" >> auth.toml
+      {%- endif %}
+      DOCKER_BUILDKIT=1 docker build \
+        --cache-from "${DOCKER_BASE_IMAGE:-ci}" \
```

Comment:
- Should `ci` be `deps` here?
- Does this actually speed anything up if the value is `ci` or `deps`?
- When/where would we set `DOCKER_BASE_IMAGE`?

Reply:
This article shows that of we want to benefit from the Docker cache, we need to build and push all stages: https://pythonspeed.com/articles/faster-multi-stage-builds/

Reply:
I think that blog post applies to multi-stage builds where a stage is build-only and "discarded" (its layers are not included in the final image). But for this Dockerfile the layers are "sequential" - `deps` contains `base`, `ci` and `app` contains `deps`, `dev` contains `ci`.

Reply:
I fixed the issue with DOCKER_BASE_IMAGE - that was indeed not set (by accident), and no speedup would result (`ci` or `deps` weren't cached of course).

Reply:
OK, got it, thanks!